File: longest_substring_without_repeating _Characters/longestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        if len(s) <= 1: return len(s)

        i, j = 0, 1
        max_len = 1
        
        while i < len(s) and j < len(s):
            if s[j] not in s[i: j]:
                j += 1
            else:
                i += 1
            max_len = max([max_len, len(s[i: j])])
        
        return max_len

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.length() <= 1) return s.length();

        unordered_set<char> st;
        int l = 0, i;
        int max_len = 1;

        for(i = 0; i < s.length(); i ++) {
            if (i == l || st.find(s[i]) == st.end()) st.insert(s[i]);
            else {
                max_len = max(max_len, i - l);
                st.erase(s[l]);
                l ++;
                i --;
            }
        }

        
        return max(max_len, i - l);
    }
};